feat(drawing): add missing connection styles in `draw_networkx_edge_labels` and `display`

Peiffap opened the pull request:
Fixes #8106.

This contains the suggested fix by @dg-pb in #8106 to allow `"arc"` and `"angle"` in `draw_networkx_edge_labels`. I'm not sure the `AssertionError` for invalid `pos` is sufficiently clear. I've also augmented the test introduced in #7739 to check all styles.

On the (slightly modified) example provided in the linked issue, we now get:
```python
import matplotlib.pyplot as plt
import networkx as nx

g = nx.DiGraph()
g.add_edge(1, 2)

pos = nx.spring_layout(g)

nx.draw_networkx_nodes(g, pos)
nx.draw_networkx_labels(g, pos)

edge = (1, 2)

styles = ['arc', 'angle']

for cs in styles:
    nx.draw_networkx_edges(g, pos, edgelist = [edge], connectionstyle = cs)
    nx.draw_networkx_edge_labels(g, pos, edge_labels = {edge: 'edge'}, connectionstyle = cs)

plt.show()
```
![cs](https://github.com/user-attachments/assets/40004f30-43ae-4be6-b0e3-27a8e8a432f6)

dg-pb commented:
> I'm not sure the `AssertionError` for invalid `pos` is sufficiently clear.

I added it as a precaution thinking that some higher level assertion already exists.
But it doesn't.
And maybe it shouldn't as values below 0 and above 1 are most likely valid.
They should just extrapolate beyond bounds.

Try removing the assertion altogether and try values -0.1, 1.1 and see if all works as expected.

Peiffap commented:
I meant the check on the length of `points`, although maybe that's just my lack of familiarity with this part of the codebase showing. I'll play around with values outside of [0, 1] as well and see what happens then. Perhaps a warning makes sense if we're relying on extrapolation?

dg-pb commented:
> I meant the check on the length of `points`

That is adaptation to `matplotlib`. Given its current state it is exact and IMO unlikely to change.

You can check those classes in `import matplotlib.patches` - it can be seen very clearly what numbers of vertices are possible for each and what sort of bezier curve it is constructing.

By the way, maybe renaming `points` to `vertices` would be a good idea - just to keep variable names consistent.



> Perhaps a warning makes sense if we're relying on extrapolation?

If extrapolation works, then I don't think a warning is needed.
If this is allowed and considered a feature then it would be very annoying to get warnings if one decides to make use of it.

So I say either restrict it at higher level or leave it as it is without errors or warnings if it works well.

dg-pb commented:
Yes, it wasn't correctly picking it up. 3 edits above fix it if that is preferred.

Peiffap commented:
Thanks for being patient with me! I'm not very familiar with the drawing functionality so I apologize for the additional burden I'm placing on you. I've added your suggestions to get rid of the assertion, and added tests to ensure it works correctly.
I've also amended the relevant commits to add you as co-author (sorry for skipping this initially, I was in a rush earlier!).

mjschwenne commented:
We recently introduced a new visualization API in #7589, and it would be much appreciated if any improvements to the old visualization system were also reflected there. Long term, we'd like to shift focus to the new visualization method, `display`. We will keep the old functions but I'd like to eventually have them be wrappers of `display` so that we aren't maintaining two separate ways to do the same thing.

dg-pb commented:
It would be great to avoid keeping 2 copies of such a thing.

We could place it outside those functions as a base, then Inherit within functions:
```python
class CurvedArrowTextBase:
    # All code here
    ...

def foo():
    class CurvedArrowText(CurvedArrowTextBase, mpl.text.Text)
        pass

def bar():
    class CurvedArrowText(CurvedArrowTextBase, mpl.text.Text)
        pass
```

dg-pb commented:
> I'm not 100% sure we're covering all the cases

I think I have covered all of them.

Angle has 2 possibilities:
1. [MOVETO, LINETO, LINETO]
2. [MOVETO, LINETO, CURVE3, CURVE3, LINETO]

Arc has 3:
1. [MOVETO, LINETO]
2. [MOVETO, LINETO, CURVE3, CURVE3, LINETO] (this one has 2 variations, but the treatment of them both is the same)
3. [MOVETO, LINETO, CURVE3, CURVE3, LINETO, CURVE3, CURVE3, LINETO]

We are interested in `[ANY, LINETO]`, which I think I extract for each case correctly.
These are lines, the rest is edge rounding, which I do not take into account (that would just be extra 100% of complexity for 2% benefit).

Then, all that is left is to identify in which line the requested position resides.

---

I think this is good to go and with this pretty much 99% of connection styles are covered (except label adaptation to minuscule curvature of rounded edges).

---

I would just like to reiterate myself that I think it would be of benefit to factor out `CurvedArrowText` class as I suggested in https://github.com/networkx/networkx/pull/8108#issuecomment-2968295897 to avoid unnecessary repetition of pretty hefty code. If not in this PR, then at least promptly after.

rossbar commented on networkx/drawing/tests/test_pylab.py:
```diff
@@ -1515,26 +1515,43 @@ def test_hide_ticks(method, hide_ticks, subplots):
     edge = (0, 1)
     G = nx.DiGraph([edge])
-    pos = {n: (n, 0) for n in G}  # Edge is horizontal line between (0, 0) and (1, 0)
+    pos = nx.spring_layout(G, k=1, seed=42)
```
Is this change necesssary? When I change this back I get assertion errors, but I don't see why the original position is invalid.

dg-pb replied:
Could you provide a minimal example where it happens? I can look into it.

dg-pb replied:
Ok, the issue is:

>           If given, the arrow curve is clipped by *patchA* and *patchB*,
>           allowing it to start/end at the border of these patches.
>           Additionally, the arrow curve can be shortened by *shrinkA* and *shrinkB*
>           to create a margin between start/end (after possible clipping) and the
>           drawn arrow.

Thus, there are 2 possible subsequent truncations of the `Path`.
They do not happen often, but can happen.
Given there were no problems with `Arc3` and `Angle3`, hopefully curves are not affected, which is most likely as all points are needed to infer any part of the path.

However, this does affect composite path and can remove points.
Luckily, all the rest of connection styles are line-based, thus there is a general solution, which might do the trick nicely. And it does simplify the code a bit:

```python
        def _update_text_pos_angle(self, arrow):
            ...

            # 1. Calculate x and y
            points = path_disp.vertices
            if isinstance(conn, (mpl.patches.ConnectionStyle.Arc3,
                                 mpl.patches.ConnectionStyle.Angle3)):
                # Arc3 or Angle3 type Connection Styles - Bezier curve
                is_line = False
                (x1, y1), (cx, cy), (x2, y2) = points
                x = tt**2 * x1 + 2 * t * tt * cx + t**2 * x2
                y = tt**2 * y1 + 2 * t * tt * cy + t**2 * y2
            else:
                is_line = True
                assert isinstance(conn, (
                    mpl.patches.ConnectionStyle.Bar,
                    mpl.patches.ConnectionStyle.Angle,
                    mpl.patches.ConnectionStyle.Arc
                ))
                # A. Collect lines
                lines = []
                codes = path_disp.codes
                LINETO =  mpath.Path.LINETO
                for i in range(1, len(points)):
                    if codes[i] == LINETO:
                        lines.append(points[i-1:i+1])
                # B. If more than one line, find the right one and position in it
                nlines = len(lines)
                if nlines != 1:
                    dists = [math.dist(*line) for line in lines]
                    dist_tot = sum(dists)
                    cdist = 0
                    last_cut = 0
                    i_last = nlines - 1
                    for i, dist in enumerate(dists):
                        cdist += dist
                        cut = cdist / dist_tot
                        if i == i_last or t < cut:
                            t = (t - last_cut) / (dist / dist_tot)
                            tt = 1 - t
                            lines = [lines[i]]
                            break
                        last_cut = cut
                [[(cx1, cy1), (cx2, cy2)]] = lines
                x = cx1 * tt + cx2 * t
                y = cy1 * tt + cy2 * t

            ...
```

dg-pb replied:
Ran this with:
1. The issue example code
2. This PR example code
3. breaking test case

All seem to be working fine

rossbar reviewed (approved):
This LGTM in so far as I follow it :). Ultimately it'd be great to see if some of this code wouldn't be a nice candidate for upstreaming to `matplotlib`. I'm sure NX isn't the only library that would benefit from having a method on the FancyArrowPatches that allows users to get the position/angle at an arbitrary distance along the curve!

That's a pie-in-the-sky goal though - for now this looks to me like it will add the missing feature for the addditional connectionstyles and is no more complex than what we already have. Thanks @dg-pb and @Peiffap !

dg-pb commented:
> Ultimately it'd be great to see if some of this code wouldn't be a nice candidate for upstreaming to `matplotlib`.

If that is needed and someone is willing, then I think this contains big part of components and logic for some generic solution:
1. Segment any path to its constituent components. Currently this is done only for LINETO segments. Would need to incorporate CURVE3 and CURVE4, but that is easy.
2. Calculate lengths of all of those. Currently this is done only for LINETO. CURVE3 and CURVE4 are missing.
3. Identify which segment desired point lies in. Logic is there.
4. Calculate the position and angle on the segment of interest. Logic is there for CURVE3 and LINETO. CURVE4 is missing.

---
---

Thanks everyone! Was a pleasure. (As always is in this repo.)

Peiffap commented:
I was actually referring to what Matt said about moving from the `draw` API to `display`, but I realize now that wasn't very clear.

Regardless... I'm not sure if what you suggested is the neatest way to do things, but it's a very simple change indeed and a definite step in the right direction, so in it goes! Thanks @dg-pb :)

rossbar commented on networkx/drawing/nx_pylab.py:
```diff
@@ -663,93 +806,8 @@ def build_fancy_arrow(e):
-            self.set_rotation(self.angle)
-            # redraw text
-            super().draw(renderer)
+    class CurvedArrowText(CurvedArrowTextBase, mpl.text.Text):
+        pass
```
Thanks for the clarification - this wasn't immediately obvious to me from the diffs alone. Just wanted to make sure we weren't abstracting for abstraction's sake!
